### ISSUE TITLE
Adding Firefox support via Puppeteer with the WebDriver BiDi protocol

### ISSUE THREAD

**Issue by whimboo** [browserless/browserless#4117]
Given that browserless supports both Playwright and Puppeteer, it would be fantastic if Firefox support could be integrated when using Puppeteer. For some time, Puppeteer has offered experimental support for Firefox through an experimental CDP implementation, but as the WebDriver BiDi implementation progresses it will soon switch to WebDriver BiDi for Firefox by default.

It's important to note that Firefox's support for APIs in Puppeteer is now at a point where website tests or scraping tasks can be executed without any modifications to existing Puppeteer scripts. The only requirement is to set the internal protocol from `cdp` to `webDriverBiDi`:

```javascript
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch({
  product: 'firefox', // or 'chrome'
  protocol: 'webDriverBiDi',
});
const page = await browser.newPage();
...
await browser.close();
```

More details can be found in the [Puppeteer WebDriver BiDi Documentation](https://pptr.dev/webdriver-bidi).

Could you consider adding this support to enable a cross-browser protocol solution, as specified by the W3C, for running tasks?

Thanks!
Henrik

**Comment by joelgriffith** [comment]
Yup, this is definitely on our radar. We have a fair bit of tech established wrt to CDP based protocol we have today, but will definitely be looking into this further!

**Comment by whimboo** [comment]
That's great! Our team would be happy to assist especially if there are some CDP APIs that are currently in use which do not yet have a WebDriver BiDi equivalent. Is there an easy way to check which APIs might be missing, or is it all up to the consumer in terms of which tests are actually run?

**Comment by joelgriffith** [comment]
Thanks for the assist! I think the biggest omission, from what I can tell, is that there isn't an option to connect with BiDi yet in Puppeteer? https://pptr.dev/api/puppeteer.connectoptions, I don't see an option for protocol unfortunately -- so likely we'll need to add a parameter in connect to support that?

**Comment by whimboo** [comment]
> Thanks for the assist! I think the biggest omission, from what I can tell, is that there isn't an option to connect with BiDi yet in Puppeteer? https://pptr.dev/api/puppeteer.connectoptions, I don't see an option for protocol unfortunately -- so likely we'll need to add a parameter in connect to support that?

@orkon can you please take a look?

**Comment by OrKoN** [comment]
ConnectOptions extends [BrowserConnectOptions](https://pptr.dev/api/puppeteer.browserconnectoptions) which contains protocol. Note that Firefox only allows a single WebDriver BiDi session currently so it is not possible to connect multiple times.

**Comment by whimboo** [comment]
Also please note that Firefox needs to be launched with WebDriver BiDi enabled. So depending on when you start Firefox the preference `remote.active-protocols` needs to be already set correctly. There is no way to enable BiDi at a later time.

**Comment by joelgriffith** [comment]
I see, thanks all, we'll get started on this and see if we run into any blockers!

**Comment by whimboo** [comment]
> I see, thanks all, we'll get started on this and see if we run into any blockers!

Hi Joel, it's been a while since the last update on this issue. I wanted to check back with you if you are still interested in adding Firefox support by using the WebDriver BiDi protocol. If yes and you have a list of blockers we would appreciate your feedback so that we can plan accordingly. Thanks!

**Comment by joelgriffith** [comment]
@whimboo yes definitely! We've been busy with a lot of different objectives but it's still on my mind. Hoping we can get this in soon, likely early 2025

**Comment by whimboo** [comment]
That sounds great. And thanks for the update!
